ddtrace/llmobs/_integrations/langchain.py

```diff
@@ -109,6 +174,181 @@ def _llmobs_set_tags(
+        """
+        parent_links = parent_span._get_ctx_item(SPAN_LINKS) or []
+        pop_indices = self._get_popped_span_link_indices(parent_span, parent_links, invoker_spans, from_output)
+
+        self._set_span_links(
+            span=parent_span,
+            from_spans=[span],
+            link_from="output",
+            link_to="output",
+            popped_span_link_indices=pop_indices,
+        )
+
+    def _get_popped_span_link_indices(
+        self, parent_span: Span, parent_links: List[Dict[str, Any]], invoker_spans: List[Span], from_output: bool
+    ) -> List[int]:
+        """
+        Returns a list of indices to pop from the parent span links list
+        This is determined by if the parent span represents a chain, and if there are steps before the step
+        represented by the span that need to be removed.
+
+        This is a temporary stopgap until we trace virtually every step in the chain, and we know the last
+        step will be the last one traced.
+        """
+        parent_instance = self._instances.get(parent_span)
+        if not parent_instance or not from_output:
+            return []
+
+        parent_instance = _extract_instance(parent_instance)
+        if not _is_chain_instance(parent_instance):
+            return []
+
+        invoker_span_ids = [str(span.span_id) for span in invoker_spans]
+        return [i for i, link in enumerate(parent_links) if link["span_id"] in invoker_span_ids]
+
+    def _set_span_links(
+        self,
+        span: Span,
+        from_spans: List[Span],
+        link_from: str,
+        link_to: str,
+        popped_span_link_indices: Optional[List[int]] = None,
+    ) -> None:
+        """Sets the span links on the given span along with the existing links."""
+        existing_links = span._get_ctx_item(SPAN_LINKS) or []
+
+        if popped_span_link_indices:
+            existing_links = [link for i, link in enumerate(existing_links) if i not in popped_span_link_indices]
+
+        links = [
+            {
+                "trace_id": "{:x}".format(from_span.trace_id),
+                "span_id": str(from_span.span_id),
+                "attributes": {"from": link_from, "to": link_to},
+            }
+            for from_span in from_spans
+            if from_span is not None
+        ]
+
+        if links:
+            span._set_ctx_item(SPAN_LINKS, existing_links + links)
+
+    def _clear_instance_recordings(self, instance: Any, parent_span: Optional[Span]) -> None:
+        """
+        Deletes the references of steps in a chain from the instance id to span mapping.
+
+        The relevant instances will be recorded again if they are re-used in another chain when
+        the other chain is invoked.
+
+        We attempt to remove the current instance as well if it has no parent or its parent instance is not a chain.
+        """
+        if not _is_chain_instance(instance):
+            del self._spans[id(instance)]
```

Comment:
mayyyybe wanna safeguard this by making sure the id is in the dict before attempting to delete in case we try to delete the reference twice. Would be worth logging a warning to let us know it's happening

Reply:
made it a debug log since it could be noisy for steps we just don't trace yet (ie output parsers)